L1Trigger/Phase2L1Taus/plugins/L1HPSPFTauProducer.cc

```diff
@@ -0,0 +1,217 @@
+        }
+      }
+    }
+  }
+
+  if (useJetSeeds_) {
+    edm::Handle<std::vector<reco::CaloJet>> l1Jets;
+    evt.getByToken(tokenL1Jets_, l1Jets);
+
+    size_t numL1Jets = l1Jets->size();
+    for (size_t idxL1Jet = 0; idxL1Jet < numL1Jets; ++idxL1Jet) {
+      reco::CaloJetRef l1Jet(l1Jets, idxL1Jet);
+      if (l1Jet->pt() > minSeedJetPt_ && std::fabs(l1Jet->eta()) < maxSeedJetEta_) {
+        tauBuilder_->reset();
+        tauBuilder_->setL1PFCandProductID(l1PFCands.id());
+        tauBuilder_->setVertex(primaryVertex);
+        //tauBuilder_->setL1PFTauSeed(l1Jet);
+        tauBuilder_->setL1PFTauSeed(l1Jet, selectedL1PFCandsSignalQualityCuts);
+        tauBuilder_->addL1PFCandidates(selectedL1PFCandsSignalOrIsolationQualityCuts);
+        tauBuilder_->buildL1PFTau();
+        l1t::L1HPSPFTau l1PFTau = tauBuilder_->getL1PFTau();
+        if (l1PFTau.pt() > 1.)
+          l1PFTauCollectionUncleaned.push_back(l1PFTau);
+      }
+    }
+  }
+
+  // sort PFTau candidate collection by decreasing pT
+  std::sort(l1PFTauCollectionUncleaned.begin(), l1PFTauCollectionUncleaned.end(), isHigherPt_pfTau);
+
+  if (debug_) {
+    std::cout << "BEFORE cleaning:" << std::endl;
+    for (size_t idx = 0; idx < l1PFTauCollectionUncleaned.size(); ++idx) {
+      const l1t::L1HPSPFTau& l1PFTau = l1PFTauCollectionUncleaned.at(idx);
+      std::cout << "L1HPSPFTau #" << idx << ": " << l1PFTau;
+    }
+  }
+
+  for (auto l1PFTau : l1PFTauCollectionUncleaned) {
+    if (applyPreselection_ &&
+        !(l1PFTau.pt() > minPFTauPt_ && std::fabs(l1PFTau.eta()) < maxPFTauEta_ &&
+          l1PFTau.leadChargedPFCand().isNonnull() && l1PFTau.leadChargedPFCand()->pt() > minLeadChargedPFCandPt_ &&
+          std::fabs(l1PFTau.leadChargedPFCand()->eta()) < maxLeadChargedPFCandEta_ &&
+          (srcL1Vertices_.label() == "" ||
+           (primaryVertex.isNonnull() && l1PFTau.leadChargedPFCand()->pfTrack().isNonnull() &&
+            std::fabs(l1PFTau.leadChargedPFCand()->pfTrack()->vertex().z() - primaryVertex->zvertex()) <
+                maxLeadChargedPFCandDz_)) &&
+          l1PFTau.sumChargedIso() < maxChargedIso_ && l1PFTau.sumChargedIso() < maxChargedRelIso_ * l1PFTau.pt()))
+      continue;
+
+    bool isOverlap = false;
+    for (auto l1PFTau2 : *l1PFTauCollectionCleaned) {
+      double deltaEta = l1PFTau.eta() - l1PFTau2.eta();
+      double deltaPhi = l1PFTau.phi() - l1PFTau2.phi();
+      if ((deltaEta * deltaEta + deltaPhi * deltaPhi) < deltaR2Cleaning_) {
+        isOverlap = true;
+      }
+    }
+    if (!isOverlap) {
+      l1PFTauCollectionCleaned->push_back(l1PFTau);
+    }
+  }
+
+  if (debug_) {
+    std::cout << "AFTER cleaning:" << std::endl;
+    for (size_t idx = 0; idx < l1PFTauCollectionCleaned->size(); ++idx) {
+      const l1t::L1HPSPFTau& l1PFTau = l1PFTauCollectionCleaned->at(idx);
+      std::cout << "L1HPSPFTau #" << idx << ": " << l1PFTau;
+    }
+  }
+
+  evt.put(std::move(l1PFTauCollectionCleaned));
+}
+
```

Comment:
Please add a `fillDescriptions()` method.
Once can be autogenerated from the `L1Trigger/Phase2L1Taus/python/L1HPSPFTauProducerPF_cfi.py` file below:
```bash
edmPythonConfigToCppValidation L1Trigger/Phase2L1Taus/python/L1HPSPFTauProducerPF_cfi.py
```
resulting in
```suggestion

#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"

void
L1HPSPFTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
  // L1HPSPFTauProducerPF
  edm::ParameterSetDescription desc;
  desc.add<bool>("useJetSeeds", true);
  desc.add<double>("minPFTauPt", 20.0);
  desc.add<double>("minSignalConeSize", 0.05);
  {
    edm::ParameterSetDescription psd0;
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("neutralHadron", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("maxDz", 0.4);
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("muon", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("maxDz", 0.4);
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("electron", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("photon", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("maxDz", 0.4);
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("chargedHadron", psd1);
    }
    desc.add<edm::ParameterSetDescription>("isolationQualityCuts", psd0);
  }
  desc.add<double>("stripSizePhi", 0.2);
  desc.add<double>("minSeedJetPt", 30.0);
  desc.add<double>("maxChargedRelIso", 1.0);
  desc.add<double>("minSeedChargedPFCandPt", 5.0);
  desc.add<edm::InputTag>("srcL1PFCands", edm::InputTag("l1pfCandidates","PF"));
  desc.add<double>("stripSizeEta", 0.05);
  desc.add<double>("maxLeadChargedPFCandEta", 2.4);
  desc.add<double>("deltaRCleaning", 0.4);
  desc.add<bool>("useStrips", true);
  desc.add<double>("maxSeedChargedPFCandDz", 1000.0);
  desc.add<double>("minLeadChargedPFCandPt", 1.0);
  desc.add<double>("maxSeedChargedPFCandEta", 2.4);
  desc.add<bool>("applyPreselection", false);
  desc.add<double>("isolationConeSize", 0.4);
  desc.add<edm::InputTag>("srcL1Vertices", edm::InputTag("L1TkPrimaryVertex"));
  desc.add<double>("maxChargedIso", 1000.0);
  {
    edm::ParameterSetDescription psd0;
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("neutralHadron", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("maxDz", 0.4);
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("muon", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("maxDz", 0.4);
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("electron", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("photon", psd1);
    }
    {
      edm::ParameterSetDescription psd1;
      psd1.add<double>("maxDz", 0.4);
      psd1.add<double>("minPt", 0.0);
      psd0.add<edm::ParameterSetDescription>("chargedHadron", psd1);
    }
    desc.add<edm::ParameterSetDescription>("signalQualityCuts", psd0);
  }
  desc.add<bool>("useChargedPFCandSeeds", true);
  desc.add<double>("maxLeadChargedPFCandDz", 1000.0);
  desc.add<double>("maxSeedJetEta", 2.4);
  desc.add<std::string>("signalConeSize", "2.8/max(1., pt)");
  desc.add<edm::InputTag>("srcL1Jets", edm::InputTag("Phase1L1TJetProducer","UncalibratedPhase1L1TJetFromPfCandidates"));
  desc.addUntracked<bool>("debug", false);
  desc.add<double>("maxPFTauEta", 2.4);
  desc.add<double>("maxSignalConeSize", 0.1);

  descriptions.addWithDefaultLabel(desc);
}

```

The parameters in the `fillDescriptions()` method should be the most common defaults, while individual customisations should still be done in the python files.